File: python/grokking-algorithms/01/binarysearch.py
from typing import List


def binalysearch(numbers: List[int], target: int):
    """Search `target` from `numbers`

    Arguments:
        numbers {List[int]} -- **sorted** numbers
        target {int} -- a number to find

    Returns:
        number -- the indef of number if it's found. If not, -1.
    """
    low = 0
    high = len(numbers) - 1

    while low <= high:
        mid = (high + low) // 2
        guess = numbers[mid]

        if guess == target:
            return mid
        if guess < target:
            low = mid + 1
        else:
            high = mid - 1

    return -1

File: typescript/grokking-algorithms/01/binarysearch.ts
export function binalysearch(numbers: number[], subject: number): number {
  let lo = 0
  let hi = numbers.length - 1
  while (lo <= hi) {
    let mid = Math.floor((lo + hi) / 2)
    let guess = numbers[mid]

    if (guess === subject) {
      return mid
    }

    if (guess < subject) {
      lo = mid + 1
    } else {
      hi = mid - 1
    }
  }

  return -1
}
